File: apps/desktop-tauri/src/lib/censor.ts
// Catálogo de enmascaramiento para el modo censura (AC-1/AC-2).
//
// Una sola fuente de verdad para la lista de campos sensibles y las funciones
// de texto de mascara. La consume:
//   - `<MaskedValue>` (UI, blur + reveal-on-hover) via `isSensitive`/`maskValue`.
//   - Los exports CSV/JSON: el backend Rust enmascara en la frontera de
//     serializacion (`src-tauri/src/commands.rs` export_devices_cmd /
//     export_services_cmd) con un catalogo espejo que **DEBE MANTENERSE
//     SINCRONIZADO** con este archivo (mismos nombres de campo y mismo formato
//     de mascara parcial). Ver el comentario cruzado en el lado Rust.
//
// Estrategia por campo (spec):
//   - MAC (primary_mac/mac/gateway_mac): placeholder constante "••••", nunca
//     revelable (AC-5).
//   - IP/hostname/display_name/gateway_ip/dns_servers/cidr: mascara parcial
//     (e.g. "192.168.1.42" -> "192.168.*.*") en exports; en UI se usa blur.
//   - vendor/manufacturer/banner/product/version/port/notes: visibles.

/** Placeholder constante para MAC: nunca revela (AC-5). */
export function maskMac(): string {
    return "••••";
}

/**
 * Mascara parcial de IPv4/IPv6.
 * IPv4: cero los dos ultimos octetos -> "192.168.1.42" => "192.168.*.*".
 * IPv6: cero los ultimos 3 grupos -> "2001:db8::1:2:3" => "2001:db8::*:*:*".
 */
export function maskIp(ip: string): string {
    if (ip.includes(":")) {
        const groups = ip.split(":");
        const n = groups.length;
        if (n > 3) {
            return [...groups.slice(0, n - 3), "*", "*", "*"].join(":");
        }
        return "*".repeat(n).split("").join(":");
    }
    const octets = ip.split(".");
    if (octets.length === 4) {
        return [octets[0], octets[1], "*", "*"].join(".");
    }
    return "*";
}

/**
 * Mascara parcial de hostname: conserva la primera etiqueta, enmascara el
 * resto -> "router.lan" => "router.*". Si no hay punto, "*" completa.
 */
export function maskHostname(h: string): string {
    const idx = h.indexOf(".");
    if (idx === -1) return "*";
    return `${h.slice(0, idx)}.*`;
}

/**
 * Mascara de CIDR: enmascara la parte de direccion pero conserva el prefix-len
 * (es estructural, no un identificador) -> "192.168.1.0/24" => "192.168.*.0/24".
 */
export function maskCidr(cidr: string): string {
    const slash = cidr.indexOf("/");
    if (slash === -1) return maskIp(cidr);
    const addr = cidr.slice(0, slash);
    const prefix = cidr.slice(slash);
    return `${maskIp(addr)}${prefix}`;
}

/**
 * Mascara una lista de DNS (servidores): enmascara cada entrada como IP.
 */
export function maskDns(servers: string[]): string {
    return servers.map((s) => maskIp(s)).join(", ");
}

/** Catálogo de campos estrictamente identificadores (sensibles). */
const SENSITIVE_FIELDS = new Set<string>([
    "primary_ip",
    "primary_mac",
    "hostname",
    "display_name",
    "gateway_ip",
    "gateway_mac",
    "dns_servers",
    "cidr",
    "ip",
    "mac",
    "device_ip",
]);

/** true si el campo es un identificador estricto que debe enmascararse. */
export function isSensitive(field: string): boolean {
    return SENSITIVE_FIELDS.has(field);
}

/** Campos MAC: placeholder constante, nunca hover-revelable (AC-5). */
const MAC_FIELDS = new Set<string>(["primary_mac", "mac", "gateway_mac"]);

export function isMacField(field: string): boolean {
    return MAC_FIELDS.has(field);
}

/**
 * Dispatch por nombre de campo a la mascara de texto adecuada.
 * Usado por los exports (parcial-mask) y por `<MaskedValue>` para su estado
 * no-revelado (el contenido bajo el blur). Mantiene un solo catalogo.
 */
export function maskValue(field: string, value: string): string {
    if (!isSensitive(field)) return value;
    switch (field) {
        case "primary_mac":
        case "mac":
        case "gateway_mac":
            return maskMac();
        case "dns_servers":
            // dns_servers llega como string plano o separado por comas.
            return maskDns(value.split(",").map((s) => s.trim()));
        case "cidr":
            return maskCidr(value);
        case "hostname":
            return maskHostname(value);
        case "display_name":
            // display_name es texto libre; mascara parcial de hostname aplica
            // razonablemente (conserva primera etiqueta).
            return maskHostname(value);
        default:
            // ip / primary_ip / gateway_ip / device_ip
            return maskIp(value);
    }
}

File: apps/desktop-tauri/src-tauri/src/censor.rs
//! Helpers de enmascaramiento para el "modo censura" (AC-1/AC-2/AC-7).
//!
//! Enmascara identificadores estrictos (IP/MAC/hostname/display_name/cidr/dns)
//! en exports para que ningún valor real salga del proceso mientras el modo
//! está activo. El catálogo de campos sensibles y el formato de cada máscara
//! DEBEN mantenerse sincronizados con `src/lib/censor.ts` (Step 4 del plan
//! `censura-mode`): misma lista de campos, mismo formato, para que UI y
//! exports no diverjan ("what you see is what you share").
//!
//! Campos sensibles (catálogo): `primary_ip`, `primary_mac`, `hostname`,
//! `display_name`, `gateway_ip`, `gateway_mac`, `dns_servers`, `cidr`, `ip`,
//! `mac`, `device_ip`. No sensibles: `vendor`, `manufacturer`, `banner`,
//! `product`, `version`, `port`, `notes`.
//!
//! Nota técnica: `Device.primary_ip` (`Option<IpAddr>`) y `primary_mac`
//! (`Option<MacAddr>`) son campos tipados que no pueden alojar cadenas
//! enmascaradas con `*`, por lo que [`mask_device`] solo muta los campos
//! `String` (`hostname`, `display_name`); IP/MAC se enmascaran en el borde de
//! serialización (JSON vía [`mask_device_value`], CSV vía registros manuales en
//! `commands.rs`). Análogamente [`mask_service_row`] muta `display_name` y
//! `device_ip` se enmascara vía [`mask_service_value`].

use mylan_core::Device;
use mylan_db::service_repo::ServiceExportRow;

/// Máscara constante para MAC (nunca se revela — AC-5). Coincide con `maskMac()`
/// en `src/lib/censor.ts`.
pub fn mask_mac() -> &'static str {
    "••••"
}

/// Enmascara una IP: IPv4 zeroa los dos últimos octetos (`a.b.*.*`); IPv6
/// zeroa los 3 últimos grupos. Coincide con `maskIp()` en `src/lib/censor.ts`.
pub fn mask_ip(s: &str) -> String {
    if s.contains('.') && !s.contains(':') {
        let parts: Vec<&str> = s.split('.').collect();
        if parts.len() == 4 {
            return format!("{}.{}.*.*", parts[0], parts[1]);
        }
    }
    if s.contains(':') {
        let parts: Vec<&str> = s.split(':').collect();
        if parts.len() > 3 {
            let head = &parts[..parts.len() - 3];
            return format!("{}:*:*:*", head.join(":"));
        }
        return "*:*:*".to_string();
    }
    s.to_string()
}

/// Enmascara un hostname: si tiene puntos conserva la primera etiqueta y
/// enmascara el resto (`router.lan` → `router.*`); si es una etiqueta única,
/// enmascara todo (`*`). Coincide con `maskHostname()` en `src/lib/censor.ts`.
pub fn mask_hostname(s: &str) -> String {
    if let Some((first, _)) = s.split_once('.') {
        format!("{first}.*")
    } else {
        "*".to_string()
    }
}

/// Enmascara un CIDR: enmascara la parte de dirección con [`mask_ip`] y
/// conserva el prefijo (`192.168.1.0/24` → `192.168.*.*/24`). Coincide con
/// `maskCidr()` en `src/lib/censor.ts`.
///
/// Sin uso actual en exports de devices/services (esas filas no incluyen
/// `cidr`), pero se mantiene como parte del catálogo compartido con
/// `src/lib/censor.ts` para que UI y backend no diverjan.
#[allow(dead_code)]
pub fn mask_cidr(s: &str) -> String {
    if let Some((addr, prefix)) = s.split_once('/') {
        format!("{}/{}", mask_ip(addr), prefix)
    } else {
        mask_ip(s)
    }
}

/// Enmascara cada servidor DNS de un vector. Coincide con `maskDns()` en
/// `src/lib/censor.ts`.
///
/// Sin uso actual en exports de devices/services, pero se mantiene como parte
/// del catálogo compartido con `src/lib/censor.ts`.
#[allow(dead_code)]
pub fn mask_dns(vec: &[String]) -> Vec<String> {
    vec.iter().map(|s| mask_ip(s)).collect()
}

/// Enmascara los campos `String` de un `Device` (`hostname`, `display_name`)
/// in situ. IP/MAC se enmascaran en el borde de serialización (ver doc del
/// módulo). Guarda cada campo con `Some(..)` antes de mutar.
pub fn mask_device(d: &mut Device) {
    if let Some(h) = d.hostname.take() {
        d.hostname = Some(mask_hostname(&h));
    }
    if let Some(n) = d.display_name.take() {
        d.display_name = Some(mask_hostname(&n));
    }
}

/// Enmascara `display_name` de una fila de servicios in situ. `device_ip`
/// (`Option<IpAddr>`) se enmascara en el borde de serialización vía
/// [`mask_service_value`].
pub fn mask_service_row(r: &mut ServiceExportRow) {
    if let Some(n) = r.display_name.take() {
        r.display_name = Some(mask_hostname(&n));
    }
}

/// Enmascara los campos IP/MAC de un `serde_json::Value` que representa un
/// `Device` serializado. Sobreescribe `primary_ip` (si es string) con
/// [`mask_ip`] y `primary_mac` con [`mask_mac`]. `hostname`/`display_name` ya
/// vienen enmascarados por [`mask_device`] sobre el struct.
pub fn mask_device_value(v: &mut serde_json::Value) {
    if let Some(obj) = v.as_object_mut() {
        if let Some(ip) = obj.get("primary_ip").and_then(|i| i.as_str()) {
            obj["primary_ip"] = serde_json::Value::String(mask_ip(ip));
        }
        if obj.contains_key("primary_mac") {
            obj["primary_mac"] = serde_json::Value::String(mask_mac().to_string());
        }
    }
}

/// Enmascara `device_ip` de un `serde_json::Value` que representa un
/// `ServiceExportRow` serializado. `display_name` ya viene enmascarado por
/// [`mask_service_row`] sobre el struct.
pub fn mask_service_value(v: &mut serde_json::Value) {
    if let Some(obj) = v.as_object_mut() {
        if let Some(ip) = obj.get("device_ip").and_then(|i| i.as_str()) {
            obj["device_ip"] = serde_json::Value::String(mask_ip(ip));
        }
    }
}

/// Serializa un valor serde a su forma de cadena CSV (replica la salida que
/// produciría `csv::Writer::serialize` sobre el campo): `null` → vacío,
/// `bool` → `"true"`/`"false"`, número → su cadena, string → el valor.
pub fn csv_str<T: serde::Serialize>(v: &T) -> String {
    match serde_json::to_value(v) {
        Ok(serde_json::Value::Null) | Err(_) => String::new(),
        Ok(serde_json::Value::Bool(b)) => b.to_string(),
        Ok(serde_json::Value::Number(n)) => n.to_string(),
        Ok(serde_json::Value::String(s)) => s,
        Ok(other) => other.to_string(),
    }
}